src/broker/content-checkpoint.c

```diff
@@ -51,45 +126,72 @@ static void checkpoint_get_continuation (flux_future_t *f, void *arg)
-                                     const flux_msg_t *msg, void *arg)
+static int content_checkpoint_get_backing (struct content_checkpoint *checkpoint,
+                                           const flux_msg_t *msg,
+                                           const char *key,
+                                           const char **errstr)
 {
-    struct content_checkpoint *checkpoint = arg;
     const char *topic = "content-backing.checkpoint-get";
-    const char *s = NULL;
-    const flux_msg_t *msgcpy = flux_msg_incref (msg);
     flux_future_t *f = NULL;
+    const flux_msg_t *msgref;
 
-    /* Temporarily maintain ENOSYS behavior */
-    if (checkpoint->rank == 0
-        && !content_cache_backing_loaded (checkpoint->cache)) {
-        errno = ENOSYS;
+    if (!(f = flux_rpc_pack (checkpoint->h, topic, 0, 0, "{s:s}", "key", key))
+        || flux_future_then (f,
+                             -1,
+                             checkpoint_get_continuation,
+                             checkpoint) < 0)
         goto error;
-    }
 
-    if (flux_request_decode (msg, NULL, &s) < 0)
+    msgref = flux_msg_incref (msg);
+    if (flux_future_aux_set (f,
+                             "msg",
+                             (void *)msgref,
+                             (flux_free_f)flux_msg_decref) < 0) {
+        flux_msg_decref (msgref);
         goto error;
+    }
 
-    if (!(f = flux_rpc (h, topic, s, 0, 0))
-        || flux_future_aux_set (f,
-                                "msg",
-                                (void *)msgcpy,
-                                (flux_free_f)flux_msg_decref) < 0
-        || flux_future_then (f,
-                             -1,
-                             checkpoint_get_continuation,
-                             checkpoint) < 0) {
-        flux_log_error (h, "error starting checkpoint-get RPC");
+    return 0;
+
+error:
+    (*errstr) = "error starting checkpoint-get RPC";
+    flux_future_destroy (f);
+    return -1;
+}
+
+void content_checkpoint_get_request (flux_t *h, flux_msg_handler_t *mh,
+                                     const flux_msg_t *msg, void *arg)
+{
+    struct content_checkpoint *checkpoint = arg;
+    const char *key;
+    const char *errstr = NULL;
+
+    if (flux_request_unpack (msg, NULL, "{s:s}", "key", &key) < 0)
         goto error;
+
+    if (checkpoint->rank == 0
+        && !content_cache_backing_loaded (checkpoint->cache)) {
+        struct checkpoint_data *data = zhashx_lookup (checkpoint->hash, key);
+        if (!data) {
+            errstr = "checkpoint key unavailable";
+            errno = ENOENT;
+            goto error;
+        }
+        if (flux_respond_pack (h, msg,
+                               "{s:O}",
+                               "value", data->value) < 0)
+            flux_log_error (h, "error responding to checkpoint-get");
+        return;
     }
 
+    if (content_checkpoint_get_backing (checkpoint, msg, key, &errstr) < 0)
+        goto error;
+
```

Comment:
For rank 0, this is forcing the RPC to use the backing store, so it would get ENOSYS in the "none" case.  The request should go to the broker RPC.  Same comment applies to `put`.

It looks like `t0028-content-backing-none.t` includes a test that expects this behavior.  Shouldn't it work?

Reply:
I hate to keep dragging this out.  Do we have a use case for _not_ returning ENOSYS to the checkpoint operations on rank > 0?    This is primarily used internally by the kvs on rank 0 only, and by the dump/restore and startlog tools, all of which one would expect  to run on rank0 I think.  Up to you but we could just prune that case and call it good if it cuts down the code.  I maybe shouldn't have brought it up, but I was thinking that we could just load the service on rank 0 and let requests be forwarded "naturally", forgetting that the service name is now shared with the content load/store ops that must work on all ranks.

Reply:
My initial thought upon reading your comments was that I screwed up.  We should have `content.checkpoint-{get,put}` forward to rank 0's `content.checkpoint-{get,put}` on non-rank 0 brokers.  But your comment makes sense that maybe we don't have to, we should just `ENOSYS` right off the bat for ranks != 0.

For myself, I tend to lean towards "consistency", b/c I just dislike seeing something different in the code.

Let me have rank > 0 forward appropriately to rank 0.  I'll spin that off into another PR since it's sort of independent of this and really a mistake from #4463